miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,587 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef ADMFEM_HPP 
+#define ADMFEM_HPP
+
+#include "mfem.hpp"
+#include "fdual.hpp"
+#include "tadvector.hpp"
+#include "taddensemat.hpp"
+
+#ifdef MFEM_USE_CODIPACK
+#include <codi.hpp>
+
+namespace mfem {
+
+namespace ad {
+#ifdef MFEM_USE_ADFORWARD
+typedef codi::RealForward ADFloatType;
+typedef TADVector<ADFloatType> ADVectorType;
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#else
+typedef codi::RealReverse ADFloatType;
+typedef TADVector<ADFloatType> ADVectorType;
+typedef TADDenseMatrix<ADFloatType> ADMatrixType;
+#endif
+}
+
+
+template<int vector_size=1, int state_size=1, int param_size=0>
+class VectorFuncAutoDiff
+{
+public:
+    VectorFuncAutoDiff(std::function<void(mfem::Vector&, ad::ADVectorType&, ad::ADVectorType&)> F_)
```

Comment:
It would be nice to have a clear description of what `F_` is supposed to do. The declaration offers few clues about the role of each parameter. `F_` is free to modify all of them, so it's unclear if they are all inout parameters or some inout and some out. It might be worth doing something like:
```C++
  /// Doxygen comments to explain the expectations on functions of this type. Which arguments are inputs?
  /// Which arguments are outputs?
  using ADFunction = std::function<void(mfem::Vector&, ad::ADVectorType&, ad::ADVectorType&)>;
```

Reply:
Description added